.github/workflows/rust.yml

```
@@ -0,0 +1,93 @@
+name: Rust
+
+on: [push, pull_request]
```

```suggestion
on: 
  push:
    branches: '*'
  pull_request:
    branches: '*'
```

This currently runs CI twice on PRs.

---

Running the PR twice is actually good because a `push` runs with the branch as it's own base and `pull_request` runs with the upstream as the base branch.

---

I'm not sure I follow on this. When you say "with the upstream as the base" do you mean "as if the PR were rebased on top of the branch being merged into"? I haven't heard anything about `push` and `pull_request` treating the same commit differently.

---

As per https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request
> Note that `GITHUB_SHA` for this event is the last merge commit of the pull request merge branch. If you want to get the commit ID for the last commit to the head branch of the pull request, use `github.event.pull_request.head.sha` instead.

Thus, the checkout action running on something triggered by `pull_request` will checkout a merge between the upstream and the PR and test it. And `push` just checks out to the commit that is pushed.

---

Hmm, interesting. I didn't know that GH did that temporary merge for `pull_request`.

Personally I generally try to avoid merging branches into a base branch which is ahead of it (e.g. through first merging back into the PR branch) to check this explicitly rather than relying on GH.